### PR TITLE
feat(quiz): unify edit-modal class picker on rosterIds (M6)

### PR DESCRIPTION
**Pull request by OPS-PIvers**
## What (M6 — backlog `docs/remaining-todos-audit.md`)

Brings the **quiz assignment edit modal** to parity with the create flow:
- `QuizAssignmentSettingsModal` now uses the unified `<AssignClassPicker>` (multi-select, Select-All, ClassLink badges, smart roster filtering) instead of the inline period-name checkbox list.
- Adds `rosterIds?: string[]` to the assignment shape (additive). The modal hydrates the picker from `rosterIds` (falling back to mapping legacy `periodNames` → rosters by name), and on save **writes both `rosterIds` and `periodNames`** via `deriveSessionTargetsFromRosters` — the same derivation the create flow uses.

**Decided scope:** additive + back-compat. `periodNames[]` is kept for read-time fallback; **no backfill** of existing assignments; the read-fallback removal is a deferred, Paul-gated cleanup.

## Critical fix included (caught by adversarial review)

The first build routed the modal's save through `updateAssignmentSettings()`, which only mirrors `periodNames` to the live session doc — leaving `sessionData.rosterIds`/`classIds`/`classId`/`classPeriodByClassId` **stale after an edit**, which breaks SSO + PIN join (`pinLoginV1` reads `sessionData.rosterIds`). Fixed by routing class targeting through the existing `setAssignmentRosters()` (the create-flow dual-write) and the remaining settings through `updateAssignmentSettings()`. See commit `396ae005`.

## Validation
- `pnpm type-check` ✓ · `pnpm build` ✓ · `vitest` ✓ (QuizAssignmentSettingsModal.test + useQuizAssignments.test — 53 tests, incl. 4 new: hydrate-from-rosterIds, hydrate-from-legacy-periodNames, write-both-on-save, clear-all).

## Reviewer notes / residual
- The 5-line `onSave` wiring in `Widget.tsx` isn't covered by an isolated unit test (Widget is heavy to render); both halves it composes — modal patch shape and `setAssignmentRosters` session dual-write — are individually tested. Integration test is a reasonable follow-up.
- Edge case: editing an assignment whose targeted roster was since **renamed/deleted** will drop that target on save (the picker can't rehydrate it). This is marginally more correct than the old verbatim period-name round-trip, but worth knowing.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: components/widgets/QuizWidget/Widget.tsx [hunk 1]
```diff
@@ -2024,7 +2024,32 @@ export const QuizWidget: React.FC<{ widget: WidgetData }> = ({ widget }) => {
...
+              const settingsPatch = { ...patch };
+              delete settingsPatch.rosterIds;
+              delete settingsPatch.periodName;
+              delete settingsPatch.periodNames;
+              if (rosterIds !== undefined) {
+                const targets = deriveSessionTargetsFromRosters(
+                  rosters.filter((r) => rosterIds.includes(r.id))
+                );
+                await setAssignmentRosters(editingAssignment.id, targets);
+              }
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/QuizWidget/Widget.tsx]: Fixed in b03bbc0 — `onSave` now builds a `Set` of the selected roster IDs and filters `rosters` with `selectedRosterIds.has(r.id)` (O(1) lookup) instead of `rosterIds.includes(r.id)`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/QuizWidget/components/QuizAssignmentSettingsModal.tsx [hunk 2]
```diff
@@ -48,24 +55,60 @@ interface SettingsOptions {
...
+ *      here — name-matching is the only legacy path.
+ *
+ * Unknown roster IDs / names simply drop out — the picker only shows rosters
+ * the teacher still has, mirroring `resolveAssignmentTargets`.
+ */
+function hydratePickerValue(
+  a: QuizAssignment,
+  rosters: ClassRoster[]
+): AssignClassPickerValue {
+  if (a.rosterIds && a.rosterIds.length > 0) {
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/QuizWidget/components/QuizAssignmentSettingsModal.tsx]: Fixed in b03bbc0. `hydratePickerValue` now gates on `a.rosterIds !== undefined`, so an explicit `[]` is treated as authoritative ("no rosters selected") and short-circuits before the legacy `periodNames` name-matching — only a genuinely absent field consults the legacy path now. Added a regression test (`an explicit empty rosterIds short-circuits the legacy periodNames fallback`) that pins this with a non-empty `periodNames` present.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/QuizWidget/Widget.tsx [hunk 3]
```diff
@@ -2024,7 +2024,32 @@ export const QuizWidget: React.FC<{ widget: WidgetData }> = ({ widget }) => {
...
+              if (Object.keys(settingsPatch).length > 0) {
+                await updateAssignmentSettings(
+                  editingAssignment.id,
+                  settingsPatch
+                );
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/QuizWidget/Widget.tsx]: Valid — the two commits aren't atomic. I went with a targeted mitigation in b03bbc0 rather than the full combined-batch hook method, and documented the residual window:

I reordered so the **secondary settings commit FIRST and the class targeting commits LAST**. Targeting (`rosterIds`/`classIds`/`classId` on the session doc, which `pinLoginV1` + SSO join read) already moves as a single all-or-nothing batch *inside* `setAssignmentRosters`. With this ordering, if the targeting commit fails the assignment keeps its **prior** targeting fully intact — students still join with the old classes — and the modal stays open for retry. The only residual inconsistency is "updated settings against old targeting," which is harmless (no broken join state). The previous order could leave new targeting live against a stale `className`/`dueAt`/`plc`.

The fully-atomic `retargetAndUpdateAssignment(id, { targets, settings })` single-batch method you sketched is the right end state — I've left a code comment marking it as the tracked follow-up, since merging it cleanly means refactoring the PLC-sweep / `deleteField` / sessionOptions-mirroring logic shared with `updateAssignmentSettings` and is worth its own change + tests.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/QuizWidget/Widget.tsx [hunk 4]
```diff
@@ -2024,7 +2024,44 @@ export const QuizWidget: React.FC<{ widget: WidgetData }> = ({ widget }) => {
...
+              // the session doc, which would leave `sessionData.rosterIds`
+              // stale and break `pinLoginV1`. Route targeting through
+              // `setAssignmentRosters` (the same dual-write the create flow
+              // uses), and send the remaining settings through
+              // `updateAssignmentSettings`.
+              const { rosterIds } = patch;
+              const settingsPatch = { ...patch };
+              delete settingsPatch.rosterIds;
+              delete settingsPatch.periodName;
+              delete settingsPatch.periodNames;
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/QuizWidget/Widget.tsx]: Fixed in the latest commit (7c16b45) — `onSave` now destructures `const { rosterIds, periodName: _periodName, periodNames: _periodNames, ...settingsPatch } = patch;` instead of spread+`delete`, so `settingsPatch` is type-narrowed to exclude the targeting fields and a future targeting field can't silently leak into `updateAssignmentSettings`. Verification: type-check ✓ · lint ✓ · format ✓.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: components/widgets/QuizWidget/Widget.tsx [hunk 5]
```diff
@@ -2024,7 +2024,44 @@ export const QuizWidget: React.FC<{ widget: WidgetData }> = ({ widget }) => {
...
+              // residual inconsistency is updated settings against old
+              // targeting, which is harmless. Folding both into one batch
+              // (a combined hook method) is a tracked follow-up.
+              if (Object.keys(settingsPatch).length > 0) {
+                await updateAssignmentSettings(
+                  editingAssignment.id,
+                  settingsPatch
+                );
+              }
+              if (rosterIds !== undefined) {
```
[the remark this answers is not in the document]
Comment (OPS-PIvers) [reply components/widgets/QuizWidget/Widget.tsx:2063]: No code change made here. The coupling is intentional and documented in the adjacent comment block: targeting (`rosterIds`/`classIds`/`classId` on the session doc, read by `pinLoginV1` + SSO join) must be dual-written via `setAssignmentRosters` on every save so an edit can't leave the session doc stale. The `if (rosterIds !== undefined)` guard is kept as a defensive boundary for any future caller that constructs a patch without touching the picker — removing it to chase "always true today" would couple correctness to the current modal's behavior. Decoupling the two batches into a single atomic `retargetAndUpdateAssignment(id, { targets, settings })` hook method (the real end-state) is already tracked as a follow-up in the thread above; that's where the guard semantics should be revisited.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by OPS-PIvers — COMMENTED** [review]
## Automated Code Review

**Date:** 2026-06-26

### Summary
Brings the quiz assignment **edit** modal to parity with the create flow by routing class targeting through `setAssignmentRosters` (the same `rosterIds`/`classIds`/`classId`/`classPeriodByClassId` dual-write), so an edit no longer leaves the live session doc stale and breaks SSO + PIN join. Additive and back-compat (`periodNames[]` kept for read-time fallback). Well-scoped and well-tested.

### Code Quality
✓ Good. The `onSave` handler now destructures the targeting fields out of the patch (fixed this run, 7c16b45) so `settingsPatch` is type-narrowed and a future targeting field can't leak into `updateAssignmentSettings`. The two-batch write is non-atomic but deliberately ordered (settings first, targeting last) so a failed targeting commit leaves prior targeting intact — documented inline, with the fully-atomic combined hook noted as a tracked follow-up.

### Widget Standards
✓ Not applicable — touches the existing QuizWidget; no new widget registered.

### Admin Config Standards
✓ Not applicable.

### Firestore & Security
✓ No new collections or rules. The dual-write is the correctness fix: it keeps `sessionData.rosterIds` consistent for `pinLoginV1` + SSO join after an edit.

### Regression Risk
Low–medium. Touches `QuizWidget/Widget.tsx`, `QuizAssignmentSettingsModal.tsx`, `useQuizAssignments.ts`. `rosterIds` is additive; legacy `periodNames` retained for the read-fallback. Known edge case (documented): editing an assignment whose targeted roster was since renamed/deleted drops that target on save.

### Missing Tests
Adequate at the unit level (53 tests incl. hydrate-from-rosterIds, legacy-periodNames fallback, write-both-on-save, clear-all). The 5-line `onSave` wiring in `Widget.tsx` isn't covered by an isolated test; both halves it composes are individually tested. Integration test is a reasonable follow-up.

### Overall
**Merge readiness:** Ready with minor notes
Solid parity fix with the critical stale-session bug already caught and fixed; only the non-atomic write window and the renamed-roster edge case remain as documented follow-ups.

---
_This review was generated automatically. All findings should be verified by a human before merging._

---
_Generated by [Claude Code](https://claude.ai/code)_
